source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2346,113 +2357,177 @@ void COpenGLDriver::SAuxContext::flushState(GL_STATE_BITS stateBits)
+                }
+                else 
+                    continue;
+
+                const uint32_t bnd = hashVal.getBindingForAttrib(attr);
+
+                if (brandNewVAO)
+                {
+                    extGlVertexArrayAttribBinding(vao, attr, bnd);
+
+                    const asset::E_FORMAT format = static_cast<asset::E_FORMAT>(hashVal.attribFormatAndComponentCount[attr]);
+
+                    if (isFloatingPointFormat(format) && getTexelOrBlockBytesize(format) == getFormatChannelCount(format) * sizeof(double))//DOUBLE
+                        extGlVertexArrayAttribLFormat(vao, attr, getFormatChannelCount(format), GL_DOUBLE, hashVal.getRelativeOffsetForAttrib(attr));
+                    else if (isFloatingPointFormat(format) || isScaledFormat(format) || isNormalizedFormat(format))//FLOATING-POINT, SCALED ("weak integer"), NORMALIZED
+                        extGlVertexArrayAttribFormat(vao, attr, isBGRALayoutFormat(format) ? GL_BGRA : getFormatChannelCount(format), formatEnumToGLenum(format), isNormalizedFormat(format) ? GL_TRUE : GL_FALSE, hashVal.getRelativeOffsetForAttrib(attr));
+                    else if (isIntegerFormat(format))//INTEGERS
+                        extGlVertexArrayAttribIFormat(vao, attr, getFormatChannelCount(format), formatEnumToGLenum(format), hashVal.getRelativeOffsetForAttrib(attr));
+
+                    if (!updatedBindings[bnd]) {
+                        extGlVertexArrayBindingDivisor(vao, bnd, hashVal.getDivisorForBinding(bnd));
+                        updatedBindings[bnd] = true;
+                    }
+                }
+
+                if (STATE_NEQ(vertexInputParams.bindings[bnd]))//this if-statement also doesnt allow GlVertexArrayVertexBuffer be called multiple times for single binding
+                {
+                    extGlVertexArrayVertexBuffer(vao, bnd, nextState.vertexInputParams.bindings[bnd].buf, nextState.vertexInputParams.bindings[bnd].offset, hashVal.getStrideForBinding(bnd));
+                    UPDATE_STATE(vertexInputParams.bindings[bnd]);
+                }
             }
-            else {
-                if (count)
-                    COpenGLExtensionHandler::extGlBindSamplers(first, count, nextState.descriptorsParams.samplers+first);
-                first = invalid_ix;
-                count = 0u;
+            if (STATE_NEQ(vertexInputParams.indexBuf))
+            {
+                extGlVertexArrayElementBuffer(vao, nextState.vertexInputParams.indexBuf);
+                UPDATE_STATE(vertexInputParams.indexBuf);
             }
         }
-        memcpy(currentState.descriptorsParams.samplers, nextState.descriptorsParams.samplers, sizeof(currentState.descriptorsParams.samplers));
+    }
+#undef STATE_NEQ
+#undef UPDATE_STATE
+    if ((stateBits & GSB_DESCRIPTOR_SETS) && currentState.pipeline)
+    {
+        GLuint ubonames[MAX_UBO_BINDING_COUNT]{};
```

Comment:
why do you even need these arrays?

Reply:
just read the code

Reply:
still explain?

Reply:
so that i can call multibind call just once and instantly also unbind anything other than current descriptor sets' descriptors

Reply:
is line 2031 your attempt at vulkan pipeline compatibility rules?

Reply:
I'd rather pay for max 4 multibinds (1 per set), rather than have a dynamic allocation here every frame.

These maximums will either be too small, or large enough to overflow a stack.

Reply:
@Crisspl this little issue needs to be resolved ASAP its screwing with overall design

Reply:
kinda resolved, but not quite, moving to #376

Reply:
resolved